### PR TITLE
Adaptive scroll margins to avoid fixed elements (closes #1057)

### LINKED ISSUE #1057: Click action misses an element if its hidden under fixed element
### Are you requesting a feature or reporting a bug?
bug

### What is the current behavior?
See: https://getbootstrap.com/examples/navbar-fixed-top/. If before click on the "View navbar docs" hyperlink scroll was at the bottom side of the page when testcafe scrolls page to the hyperlink but it leaves hidden under the top navbar and the hyperlink remains unavailable for click.

### What is the expected behavior?
TestCafe should check visibility of clicking element and scroll page to make it visible for click.

### How would you reproduce the current behavior (if this is a bug)?
1. Scroll page to bottom
2. ```await t.click(".btn")```

Tested page URL:
https://getbootstrap.com/examples/navbar-fixed-top/

### PR DESCRIPTION
**Pull request by AndreyBelym**
\cc @AlexanderMoskovkin @georgiy-abbasov

### CODE REVIEW

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by georgiy-abbasov** [comment]
try to rebase

**Comment by AndreyBelym** [comment]
FPR

File: test/functional/fixtures/regression/gh-1057/testcafe-fixtures/index-test.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+        .hover('#target2')
+        .click('#target2')
+        .hover('#target1')
+        .wait(5000)
+        .hover('#target1')
+        .click('#target1');
+
+    var targetsClicked = await t.eval(() => window.target1Clicked && window.target2Clicked);
+
+    await t.expect(targetsClicked).ok();
```
Comment (helen-dikareva) [inline test/functional/fixtures/regression/gh-1057/testcafe-fixtures/index-test.js]: Now client function can be in assertion, it can look better

**Comment by AndreyBelym** [comment]
FPR (Fixed test naming)

File: src/client/automation/playback/scroll.js [hunk 2]
```diff
@@ -176,18 +204,42 @@ export default class ScrollAutomation {
...
+    }
+
     run () {
         this._scrollElement();
 
-        return this._scrollParents();
+        return this
+            ._scrollParents()
+            .then(() => whilst(() => !this._isScrollMarginTooBig() && this._isElementHiddenByFixed(), () => {
+                if (this._isScrollMarginTooBig() || !this._isElementHiddenByFixed())
```
Comment (AlexanderMoskovkin) [inline src/client/automation/playback/scroll.js]: as I understand you check it in the `whilst` condition, why do you need to check it again?

File: src/client/automation/playback/scroll.js [hunk 3]
```diff
@@ -47,33 +58,44 @@ export default class ScrollAutomation {
         styleUtils.setScrollTop(element, top);
     }
 
-    static _getScrollToPoint (elementDimensions, { x, y }) {
+    _forceMaxScrollMargin (newMaxScrollMargin) {
```
Comment (AlexanderMoskovkin) [inline src/client/automation/playback/scroll.js]: I'm not sure that `force` is suitable word here, how you can force margin?
Force scrolling or increase margin, something like this I guess.

File: test/functional/fixtures/regression/gh-1057/testcafe-fixtures/index-test.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+fixture `gh-1057`
+    .page `http://localhost:3000/fixtures/regression/gh-1057/pages/index.html`;
+
+const targetsClicked = ClientFunction(() => window.target1Clicked && window.target2Clicked);
+
+test('gh-1057', async t => {
+    // NOTE: scrolling has issues in iOS Simulator https://github.com/DevExpress/testcafe/issues/1237
+    await t
+        .wait(5000)
```
Comment (AlexanderMoskovkin) [inline test/functional/fixtures/regression/gh-1057/testcafe-fixtures/index-test.js]: if we need wait too much on iOS Simulator, let's do it only in simulator. Use userAgent to determine it

Comment (AlexanderMoskovkin) [inline test/functional/fixtures/regression/gh-1057/testcafe-fixtures/index-test.js]: And I don't get what does this test do. Why we need wait 5000 at start of the test, then hover `target2`, then wait and hover `target2` again? If it's all for ios let's skip it for this test

**Comment by AndreyBelym** [comment]
FPR

File: src/client/automation/playback/scroll.js [hunk 5]
```diff
@@ -47,33 +57,37 @@ export default class ScrollAutomation {
         styleUtils.setScrollTop(element, top);
     }
 
-    static _getScrollToPoint (elementDimensions, { x, y }) {
-        var needForwardScrollLeft  = x >= elementDimensions.scroll.left + elementDimensions.width;
-        var needBackwardScrollLeft = x <= elementDimensions.scroll.left;
+    _changeMaxScrollMargin (newMaxScrollMargin) {
+        this.maxScrollMargin = newMaxScrollMargin;
```
Comment (helen-dikareva) [inline src/client/automation/playback/scroll.js]: better use setter pls

**Comment by AndreyBelym** [comment]
Checked it with functional tests in a small-sized window, seems to be OK.

**Comment by AlexanderMoskovkin** [comment]
cool

**Pull request merged**